Bug: test_pfc_frames_not_counted_as_rx_ok fails due to long run time and inadequate assertion

### Is it platform specific

mellanox

### Importance or Severity

Critical

### Description of the bug
The bug is introduced by https://github.com/sonic-net/sonic-mgmt/pull/25741. On some of Mellanox devices(SN2700), when there are multiple active interfaces, the test will inject PFC pause to each one of them and to each priority sequentially, which lead to ~30min test time. During this window, the RX_OK keeps increasing and exceeds the margin 2000. 

### Steps to Reproduce
Need to run this test on some slow devices or devices with multiple active interfaces(For example Mellanox SN2700). The longer the test runs, the test is more likely to fail, because the margin is fixed.

### Actual Behavior and Expected Behavior

The test verify that the RX_OK and RX_DRP are not increased greatly due to pfc injection, but use a fixed number of margin 2000, and the condition is RX_OK < margin and RX_DRP < margin. My question is, how is the margin determined? From our failure, it looks like these 3000+ RX_OK are not from PFC injection. Shall we modify the test to choose the margin wisely? Or we need to optimize the assertion logic so that it can identify that the RX_OK delta and RX_DRP delta really come from PFC injection?

### Relevant log output

```shell
AssertionError: PFC frames were counted as RX_OK or RX_DRP beyond the allowed margin of 2000 on the following interfaces [(intf, rx_ok_delta, rx_drp_delta)]: [('Ethernet40', 3233, 0), ('Ethernet56', 3231, 0), ('Ethernet64', 3166, 0), ('Ethernet68', 3158, 0), ('Ethernet72', 3159, 0), ('Ethernet76', 3162, 0), ('Ethernet80', 3157, 0), ('Ethernet84', 3156, 0), ('Ethernet88', 3159, 0), ('Ethernet92', 3169, 0), ('Ethernet96', 3164, 0), ('Ethernet100', 3169, 0), ('Ethernet104', 3168, 0), ('Ethernet108', 3157, 0), ('Ethernet112', 3164, 0), ('Ethernet116', 3166, 0), ('Ethernet120', 3159, 0), ('Ethernet124', 3169, 0)]
```

### Output of `show version`

```shell

```

### Attach files (if any)

_No response_

--- a/tests/common/helpers/pfc_counters.py
+++ b/tests/common/helpers/pfc_counters.py
@@ -9,6 +9,7 @@ counter verification without cross-feature imports.
 from tests.common.platform.device_utils import eos_to_linux_intf, nxos_to_linux_intf, sonic_to_linux_intf
 from tests.common.helpers.drop_counters.drop_counters import GET_L2_COUNTERS, get_pkt_drops
 import os
+import random
 import time
 import pytest
 import logging
@@ -308,12 +309,12 @@ def run_rx_ok_isolation_test(fanouthosts, duthost, conn_graph_facts,       # noq
               counted as normal RX packets (RX_OK) or RX drops (RX_DRP) on the
               DUT interfaces.
 
-              Each active physical interface is exercised one at a time: a
-              baseline snapshot is taken, a burst of PFC frames is sent across
-              all priorities to that port, then the port's counters are read
-              back. The per-port measurement window keeps the RX_OK/RX_DRP
-              deltas within `margin` instead of accumulating background traffic
-              across the whole port scan.
+              A single active physical interface is chosen at random and
+              exercised: a baseline snapshot is taken, a burst of PFC frames is
+              sent across all priorities to that port, then the port's counters
+              are read back. The narrow measurement window keeps the
+              RX_OK/RX_DRP deltas within `margin` instead of accumulating
+              background traffic across a full port scan.
     @param duthost: The object for interacting with DUT through ansible
     @param conn_graph_facts: Testbed topology connectivity information
     @param leaf_fanouts: Leaf fanout switches
@@ -347,43 +348,43 @@ def run_rx_ok_isolation_test(fanouthosts, duthost, conn_graph_facts,       # noq
             "topology and fanout connectivity."
         )
 
-    """ Send frames and read back one port at a time so background traffic
-        cannot accumulate across the whole port scan """
-    failures = []
-    for intf in active_phy_intfs:
-        peer_device = conn_facts[intf]['peerdevice']
-        peer_port = conn_facts[intf]['peerport']
-        peerdev_ans = fanouthosts[peer_device]
-        peer_port_name, fanout_hwsku = _resolve_peer_port_name(
-            peerdev_ans, enum_fanout_graph_facts, peer_port)
+    """ Exercise a single randomly-chosen port so background traffic cannot
+        accumulate across a full port scan """
+    intf = random.choice(active_phy_intfs)
+    peer_device = conn_facts[intf]['peerdevice']
+    peer_port = conn_facts[intf]['peerport']
+    peerdev_ans = fanouthosts[peer_device]
+    peer_port_name, fanout_hwsku = _resolve_peer_port_name(
+        peerdev_ans, enum_fanout_graph_facts, peer_port)
+    logger.info(
+        "Selected interface %s (peer %s port %s) out of %d candidate(s) for "
+        "PFC RX_OK isolation test", intf, peer_device, peer_port, len(active_phy_intfs))
 
-        """ Baseline for this port immediately before sending """
-        baseline = get_rx_port_counters(duthost)
+    """ Baseline for this port immediately before sending """
+    baseline = get_rx_port_counters(duthost)
 
-        for priority in range(PRIO_COUNT):
-            send_pfc_frame(peerdev_ans, peer_port_name, fanout_hwsku,
-                           priority, pause_time, pkt_count)
+    for priority in range(PRIO_COUNT):
+        send_pfc_frame(peerdev_ans, peer_port_name, fanout_hwsku,
+                       priority, pause_time, pkt_count)
 
-        """ SONiC takes some time to update counters in database """
-        time.sleep(5)
-        after = get_rx_port_counters(duthost)
+    """ SONiC takes some time to update counters in database """
+    time.sleep(5)
+    after = get_rx_port_counters(duthost)
 
-        if intf not in baseline or intf not in after:
-            logger.warning(
-                "Interface %s missing from the %s counter snapshot; skipping its "
-                "RX_OK/RX_DRP validation", intf,
-                "baseline" if intf not in baseline else "post-send")
-            continue
-        rx_ok_delta = after[intf]['RX_OK'] - baseline[intf]['RX_OK']
-        rx_drp_delta = after[intf]['RX_DRP'] - baseline[intf]['RX_DRP']
-        if rx_ok_delta > margin or rx_drp_delta > margin:
-            failures.append((intf, rx_ok_delta, rx_drp_delta))
-            logger.error(
-                "Interface %s: RX_OK increased by %d, RX_DRP increased by %d "
-                "(allowed margin %d) after receiving %d PFC frames per priority",
-                intf, rx_ok_delta, rx_drp_delta, margin, pkt_count)
+    assert intf in baseline and intf in after, (
+        "Interface {} missing from the {} counter snapshot; cannot validate "
+        "its RX_OK/RX_DRP counters"
+    ).format(intf, "baseline" if intf not in baseline else "post-send")
 
-    assert len(failures) == 0, (
+    rx_ok_delta = after[intf]['RX_OK'] - baseline[intf]['RX_OK']
+    rx_drp_delta = after[intf]['RX_DRP'] - baseline[intf]['RX_DRP']
+    if rx_ok_delta > margin or rx_drp_delta > margin:
+        logger.error(
+            "Interface %s: RX_OK increased by %d, RX_DRP increased by %d "
+            "(allowed margin %d) after receiving %d PFC frames per priority",
+            intf, rx_ok_delta, rx_drp_delta, margin, pkt_count)
+
+    assert rx_ok_delta <= margin and rx_drp_delta <= margin, (
         "PFC frames were counted as RX_OK or RX_DRP beyond the allowed margin of {} "
-        "on the following interfaces [(intf, rx_ok_delta, rx_drp_delta)]: {}"
-    ).format(margin, failures)
+        "on interface {} (rx_ok_delta={}, rx_drp_delta={})"
+    ).format(margin, intf, rx_ok_delta, rx_drp_delta)

--- a/tests/common/helpers/pfc_counters.py
+++ b/tests/common/helpers/pfc_counters.py
@@ -308,11 +308,12 @@ def run_rx_ok_isolation_test(fanouthosts, duthost, conn_graph_facts,       # noq
               counted as normal RX packets (RX_OK) or RX drops (RX_DRP) on the
               DUT interfaces.
 
-              A large burst of PFC frames is sent across all priorities to every
-              active physical interface first; only then is a single counter
-              snapshot compared against the baseline (one stats retrieval pass).
-              The RX_OK and RX_DRP deltas must each stay within `margin` to
-              tolerate background control-plane traffic.
+              Each active physical interface is exercised one at a time: a
+              baseline snapshot is taken, a burst of PFC frames is sent across
+              all priorities to that port, then the port's counters are read
+              back. The per-port measurement window keeps the RX_OK/RX_DRP
+              deltas within `margin` instead of accumulating background traffic
+              across the whole port scan.
     @param duthost: The object for interacting with DUT through ansible
     @param conn_graph_facts: Testbed topology connectivity information
     @param leaf_fanouts: Leaf fanout switches
@@ -329,49 +330,44 @@ def run_rx_ok_isolation_test(fanouthosts, duthost, conn_graph_facts,       # noq
 
     conn_facts = conn_graph_facts['device_conn'].get(duthost.hostname, {})
     int_status = asic.show_interface(command="status")['ansible_facts']['int_status']
-    """ We only test active physical interfaces that have connection graph entries """
+    """ We only test active physical interfaces that have connection graph entries
+        and a reachable fanout """
     active_phy_intfs = [intf for intf in int_status if
                         intf.startswith('Ethernet') and
                         int_status[intf]['admin_state'] == 'up' and
                         int_status[intf]['oper_state'] == 'up' and
-                        intf in conn_facts]
+                        intf in conn_facts and
+                        conn_facts[intf]['peerdevice'] in fanouthosts]
 
-    """ Baseline RX counters for all ports in a single retrieval """
-    baseline = get_rx_port_counters(duthost)
+    """ Skip if nothing exercisable was found """
+    if len(active_phy_intfs) == 0:
+        pytest.skip(
+            "No active physical interfaces with a reachable fanout were exercised, so "
+            "PFC RX counter isolation could not be validated. Check the testbed "
+            "topology and fanout connectivity."
+        )
 
-    """ Send all PFC frames first, across all priorities and all ports """
-    tested_intfs = []
+    """ Send frames and read back one port at a time so background traffic
+        cannot accumulate across the whole port scan """
+    failures = []
     for intf in active_phy_intfs:
         peer_device = conn_facts[intf]['peerdevice']
         peer_port = conn_facts[intf]['peerport']
-
-        if peer_device not in fanouthosts:
-            continue
-
         peerdev_ans = fanouthosts[peer_device]
         peer_port_name, fanout_hwsku = _resolve_peer_port_name(
             peerdev_ans, enum_fanout_graph_facts, peer_port)
+
+        """ Baseline for this port immediately before sending """
+        baseline = get_rx_port_counters(duthost)
+
         for priority in range(PRIO_COUNT):
             send_pfc_frame(peerdev_ans, peer_port_name, fanout_hwsku,
                            priority, pause_time, pkt_count)
-        tested_intfs.append(intf)
 
-    """ Guard against an empty run masquerading as a pass """
-    assert len(tested_intfs) > 0, (
-        "No active physical interfaces with a reachable fanout were exercised, so "
-        "PFC RX counter isolation could not be validated. Check the testbed "
-        "topology and fanout connectivity."
-    )
+        """ SONiC takes some time to update counters in database """
+        time.sleep(5)
+        after = get_rx_port_counters(duthost)
 
-    """ SONiC takes some time to update counters in database """
-    time.sleep(5)
-
-    """ Validate RX counters in one single retrieval swoop """
-    after = get_rx_port_counters(duthost)
-
-    """ Validate every active interface, not only the ones we sent frames to """
-    failures = []
-    for intf in active_phy_intfs:
         if intf not in baseline or intf not in after:
             logger.warning(
                 "Interface %s missing from the %s counter snapshot; skipping its "

